decayedlog.go

```diff
@@ -0,0 +1,419 @@
+// to collect entries which are expired according to their stored CLTV value
+// and the current block height. DecayedLog wraps channeldb for simplicity and
+// batches writes to the database to decrease write contention.
+type DecayedLog struct {
+	started int32
+	stopped int32
+
+	dbPath string
+
+	db *channeldb.DB
+
+	notifier chainntnfs.ChainNotifier
+
+	wg   sync.WaitGroup
+	quit chan struct{}
+}
+
+func NewDecayedLog(dbPath string,
+	notifier chainntnfs.ChainNotifier) *DecayedLog {
+
+	// Use default path for log database
+	if dbPath == "" {
+		dbPath = defaultDbDirectory
+	}
+
+	return &DecayedLog{
+		dbPath:   dbPath,
+		notifier: notifier,
+		quit:     make(chan struct{}),
+	}
+}
+
+// Start opens the database we will be using to store hashed shared secrets.
+// It also starts the garbage collector in a goroutine to remove stale
+// database entries.
+func (d *DecayedLog) Start() error {
+	if !atomic.CompareAndSwapInt32(&d.started, 0, 1) {
+		return nil
+	}
+
+	// Open the channeldb for use.
+	var err error
+	if d.db, err = channeldb.Open(d.dbPath); err != nil {
+		return fmt.Errorf("Could not open channeldb: %v", err)
+	}
+
+	// Initialize the primary buckets used by the decayed log.
+	if err := d.initBuckets(); err != nil {
+		return err
+	}
+
+	// Start garbage collector.
+	if d.notifier != nil {
+		epochClient, err := d.notifier.RegisterBlockEpochNtfn()
+		if err != nil {
+			return fmt.Errorf("Unable to register for epoch "+
+				"notifications: %v", err)
+		}
+
+		d.wg.Add(1)
+		go d.garbageCollector(epochClient)
+	}
+
+	return nil
+}
+
+// initBuckets initializes the primary buckets used by the decayed log, namely
+// the shared hash bucket, and batch replay
+func (d *DecayedLog) initBuckets() error {
+	return d.db.Update(func(tx *bolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists(sharedHashBucket)
+		if err != nil {
+			return ErrDecayedLogInit
+		}
+
+		_, err = tx.CreateBucketIfNotExists(batchReplayBucket)
+		if err != nil {
+			return ErrDecayedLogInit
+		}
+
+		return nil
+	})
+}
+
+// Stop halts the garbage collector and closes channeldb.
+func (d *DecayedLog) Stop() error {
+	if !atomic.CompareAndSwapInt32(&d.stopped, 0, 1) {
+		return nil
+	}
+
+	// Stop garbage collector.
+	close(d.quit)
+
+	d.wg.Wait()
+
+	// Close channeldb.
+	d.db.Close()
+
+	return nil
+}
+
+// garbageCollector deletes entries from sharedHashBucket whose expiry height
+// has already past. This function MUST be run as a goroutine.
+func (d *DecayedLog) garbageCollector(epochClient *chainntnfs.BlockEpochEvent) {
+	defer d.wg.Done()
+	defer epochClient.Cancel()
+
+	for {
+		select {
+		case epoch, ok := <-epochClient.Epochs:
+			if !ok {
+				// Block epoch was canceled, shutting down.
+				// TODO(conner): add logging
+				return
+			}
+
+			// Perform a bout of garbage collection using the
+			// epoch's block height.
+			height := uint32(epoch.Height)
+			if err := d.gcExpiredHashes(height); err != nil {
+				// TODO(conner): add logging
```

Comment:
As for logging, let's go with a package level logger?

Reply:
Added the `sphxLog` using a log template from one of the lnd subpkgs